File: python/956.py
# Project Euler Problem 956
#
# PROBLEM DESCRIPTION:
# <p>
# The total number of prime factors of $n$, counted with multiplicity, is denoted $\Omega(n)$.<br>
# For example, $\Omega(12)=3$, counting the factor $2$ twice, and the factor $3$ once.</p>
# 
# <p>
# Define $D(n, m)$ to be the sum of all divisors $d$ of $n$ where $\Omega(d)$ is divisible by $m$. <br>
# For example, $D(24, 3)=1+8+12=21$.</p>
# 
# <p>
# The <b>superfactorial</b> of $n$, often written as $n\$$, is defined as the product of the first $n$ factorials:
# $$n\$=1!\times 2! \times\cdots\times n!$$
# The <b>superduperfactorial</b> of $n$, we write as $n\bigstar$, is defined as the product of the first $n$ superfactorials:
# $$n\bigstar=1\$ \times 2\$ \times\cdots\times n\$ $$
# </p>
# 
# <p>
# You are given $D(6\bigstar, 6)=6368195719791280$.</p>
# 
# <p>
# Find $D(1\,000\bigstar, 1\,000)$. 
# Give your answer modulo $999\,999\,001$.</p>
#

import math

def sieve(n):
    is_prime = [True]*(n+1)
    is_prime[0]=is_prime[1]=False
    for i in range(2, int(math.sqrt(n))+1):
        if is_prime[i]:
            for j in range(i*i, n+1, i):
                is_prime[j]=False
    return [i for i in range(2,n+1) if is_prime[i]]

def mod_inverse(a, m):
    return pow(a, -1, m)

def is_primitive_root(g, mod, primes_factors):
    phi = mod - 1
    for q in primes_factors:
        if pow(g, phi // q, mod) == 1:
            return False
    return True

mod = 999999001
phi = mod - 1
factors = [2,3,5,7,11,13,37]

# find primitive root
g = 2
while not is_primitive_root(g, mod, factors):
    g += 1

n = 1000
m = 1000
primes = sieve(n)
exponents = []
for pp in primes:
    e = 0
    for k in range(1, n+1):
        vk = 0
        pk = pp
        while pk <= k:
            vk += k // pk
            pk *= pp
        e += (n - k + 1) * vk
    exponents.append(e)

# now omega = g^{phi / m}
omega = pow(g, phi // m, mod)

inv_m = mod_inverse(m, mod)

sum_S = 0
for j in range(m):
    y = pow(omega, j, mod)
    prod = 1
    for idx in range(len(primes)):
        pp = primes[idx]
        e = exponents[idx]
        r = (pp * y) % mod
        if r == 1:
            geo = (e + 1) % mod
        else:
            exp_mod = (e + 1) % phi
            re_plus1 = pow(r, exp_mod, mod)
            num = (1 - re_plus1 + mod) % mod
            den = (1 - r + mod) % mod
            den_inv = mod_inverse(den, mod)
            geo = (num * den_inv) % mod
        prod = (prod * geo) % mod
    sum_S = (sum_S + prod) % mod

result = (sum_S * inv_m) % mod
print(result)

File: c/956.c
/* Project Euler 956 - D(1000*, 1000) mod 999999001
 * Uses DFT-based approach with roots of unity.
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef long long ll;
typedef __int128 lll;
#define MOD 999999001LL

ll mod_pow(ll base, ll exp, ll mod) {
    ll result = 1;
    base %= mod;
    if (base < 0) base += mod;
    while (exp > 0) {
        if (exp & 1) result = (lll)result * base % mod;
        base = (lll)base * base % mod;
        exp >>= 1;
    }
    return result;
}

ll mod_inv(ll a, ll mod) {
    return mod_pow(a, mod - 2, mod);
}

int sieve_primes(int limit, int *primes) {
    char *is_p = (char *)calloc(limit + 1, 1);
    for (int i = 2; i <= limit; i++) is_p[i] = 1;
    for (int i = 2; (ll)i*i <= limit; i++) {
        if (is_p[i]) {
            for (int j = i*i; j <= limit; j += i) is_p[j] = 0;
        }
    }
    int cnt = 0;
    for (int i = 2; i <= limit; i++) {
        if (is_p[i]) primes[cnt++] = i;
    }
    free(is_p);
    return cnt;
}

int is_primitive_root(ll g, ll mod, int *factors, int nf) {
    ll phi = mod - 1;
    for (int i = 0; i < nf; i++) {
        if (mod_pow(g, phi / factors[i], mod) == 1) return 0;
    }
    return 1;
}

int main(void) {
    int n = 1000, m = 1000;
    int phi_factors[] = {2, 3, 5, 7, 11, 13, 37};
    int nf = 7;

    /* Find primitive root of MOD */
    ll g = 2;
    while (!is_primitive_root(g, MOD, phi_factors, nf)) g++;

    /* Get primes up to n */
    int primes[200];
    int num_primes = sieve_primes(n, primes);

    /* Compute exponents of each prime in 1000-bigstar */
    ll *exponents = (ll *)malloc(num_primes * sizeof(ll));
    for (int idx = 0; idx < num_primes; idx++) {
        int pp = primes[idx];
        ll e = 0;
        for (int k = 1; k <= n; k++) {
            ll vk = 0;
            ll pk = pp;
            while (pk <= k) {
                vk += k / pk;
                pk *= pp;
            }
            e += (ll)(n - k + 1) * vk;
        }
        exponents[idx] = e;
    }

    ll phi = MOD - 1;
    ll omega = mod_pow(g, phi / m, MOD);
    ll inv_m = mod_inv(m, MOD);

    ll sum_S = 0;
    for (int j = 0; j < m; j++) {
        ll y = mod_pow(omega, j, MOD);
        ll prod = 1;
        for (int idx = 0; idx < num_primes; idx++) {
            int pp = primes[idx];
            ll e = exponents[idx];
            ll r = (lll)pp * y % MOD;
            ll geo;
            if (r == 1) {
                geo = (e + 1) % MOD;
            } else {
                ll exp_mod = (e + 1) % phi;
                ll re_plus1 = mod_pow(r, exp_mod, MOD);
                ll num = (1 - re_plus1 + MOD + MOD) % MOD;
                ll den = (1 - r + MOD + MOD) % MOD;
                ll den_inv = mod_inv(den, MOD);
                geo = (lll)num * den_inv % MOD;
            }
            prod = (lll)prod * geo % MOD;
        }
        sum_S = (sum_S + prod) % MOD;
    }

    ll result = (lll)sum_S * inv_m % MOD;
    printf("%lld\n", result);

    free(exponents);
    return 0;
}
